app/controllers/articleController.js

```diff
@@ -116,4 +116,85 @@ export default class ArticleController {
       }
     }
   }
+  /**
+   * search Method
+   *
+   * This method searches and returns all articles matching the given params
+   *
+   * @param {HTTP REQ} req web request object
+   * @param {HTTP RES} res web response object
+   * @param {function} next middleware function
+   */
+  static async search(req, res, next){
+    try {
+      const query = {};
+      var limit;
+      
+      // construct a query json based on the queries that are in the passed json
+      if (req.body.hasOwnProperty('issueNumber')===true) {
```

Comment:
Let's have a look at the inner workings of this function. First thing I notice is that some `if` statements have `===true` (which is redundant against `hasOwnProperty` [documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty) ), so that can be eliminated.

Furthermore, there are a lot of `if` statements -- more than tends to be readable from a high level, and also, the repetition is telling that there may be some room for abstraction.

Within a lot of `if` statements, we can see that you essentially do the same thing (for user email specific list searches). You iterate through the given list of users' emails, and then ascertain their mongo ids. There may already exist a method that does this (in other PRs, but if it doesn't exist yet on main you are free to build it yourself to abstract this). 

------------

The above would probably be a fine fix -- if you're looking to take this entire method implementation up a level, we can take the `resolver` principle from GraphQL. 

Instead of having a lot of if statements check the req body over and over for the search options, it might be easier to aggregate them into a pre-existing mapping between key, and a resolver function. This resolver function should be able to `resolve` to transform/not transform data onto an object, depending on what is required. Then you can iterate through all the req.body keys to `reduce` the query to the full mongodb query here. Here is some starter code, so its clear:

```js
const mapping = {
  issueNumber: (issueNumber) => issueNumber, // resolves issue number by returning the given value
  authors: getUserIdsByEmails, // resolves author emails into ids by mapping the value to a function that does this (getUserIdsByEmails is a function)
...
};

const searchOptions = Object.keys(req.body).map(key => ({
  [key]: req.body[key]
}));

const query = searchOptions.reduce((intermediateQuery, currentOption) => {
  // specific implementation is for you to figure out
}, {});

```

Reply:
This may or may not work for async resolvers, but that is just something else to test for